bin/top.ml

```diff
@@ -58,19 +70,193 @@ let term =
           let include_paths =
             Dune_rules.Lib_flags.L.toplevel_include_paths requires
           in
-          let* files = link_deps sctx requires in
-          let+ () =
-            Memo.parallel_iter files ~f:(fun file ->
-                let+ (_ : Digest.t) = Build_system.build_file file in
-                ())
-          in
-          let files_to_load =
-            List.filter files ~f:(fun p ->
-                let ext = Path.extension p in
-                ext = Ocaml.Mode.compiled_lib_ext Byte
-                || ext = Ocaml.Cm_kind.ext Cmo)
-          in
+          let+ files_to_load = files_to_load_of_requires sctx requires in
           Dune_rules.Toplevel.print_toplevel_init_file ~include_paths
-            ~files_to_load))
+            ~files_to_load ~uses:[] ~pp:None ~ppx:None))
 
 let command = (term, info)
+
+module Module = struct
+  let doc =
+    "Print a list of toplevel directives for loading a module into the topevel."
+
+  let man =
+    [ `S "DESCRIPTION"
+    ; `P doc
+    ; `P
+        "The module's source is evaluated in the toplevel without being sealed \
+         by the mli."
+    ; `P
+        {|The output of $(b,dune toplevel-init-file) should be evaluated in a toplevel
+          to make the module available there.|}
+    ; `Blocks Common.help_secs
+    ]
+
+  let info = Term.info "top-module" ~doc ~man
+
+  let module_directives sctx mod_ =
+    let ctx = Super_context.context sctx in
+    let src = Path.Build.append_source ctx.build_dir mod_ in
+    let dir = Path.Build.parent_exn src in
+    let open Memo.O in
+    let module_name =
+      let name = src |> Path.Build.basename |> Filename.chop_extension in
+      match Dune_rules.Module_name.of_string_user_error (Loc.none, name) with
+      | Ok s -> s
+      | Error e -> raise (User_error.E e)
+    in
+    let drop_rules f =
+      let+ res, _ =
+        Memo.Implicit_output.collect Dune_engine.Rules.implicit_output f
+      in
+      res
+    in
+    let* dir_contents =
+      drop_rules @@ fun () -> Dune_rules.Dir_contents.get sctx ~dir
+    in
+    let* ocaml = Dune_rules.Dir_contents.ocaml dir_contents in
+    let stanza =
+      match Dune_rules.Ml_sources.lookup_stanza_module ocaml module_name with
+      | None -> User_error.raise [ Pp.text "" ]
```

Comment:
Shouldn't we print an error here "the given module doesn't belong to any stanza"?